utils/src/channels/tracked.rs

```diff
@@ -0,0 +1,403 @@
+        while let Some(delivered) = state.pending_sequences.get(&(current_watermark + 1)) {
+            // If the next message is not delivered, we can stop
+            if !*delivered {
+                break;
+            }
+
+            // Remove the next message from the pending list
+            state.pending_sequences.remove(&(current_watermark + 1));
+            current_watermark += 1;
+            state.watermark = current_watermark;
+        }
+
+        // Update batch count (if necessary)
+        if let Some(batch) = &self._batch {
+            let count = state.batch_counts.get_mut(batch).unwrap();
+            if *count > 1 {
+                *count -= 1;
+            } else {
+                state.batch_counts.remove(batch);
+            }
+        }
+    }
+}
+
+/// A message containing data and a [Guard] that tracks delivery.
+pub struct Message<T, B: Eq + Hash + Clone> {
+    /// The data of the message.
+    pub data: T,
+    /// The [Guard] that tracks delivery.
+    ///
+    /// When no outstanding references to the guard exist, the message is considered delivered.
+    pub guard: Arc<Guard<B>>,
+}
+
+/// The state of the [Tracker].
+struct State<B> {
+    next: u64,
+    watermark: u64,
+    batch_counts: HashMap<B, usize>,
+    pending_sequences: HashMap<u64, bool>,
+}
+
+impl<B> Default for State<B> {
+    fn default() -> Self {
+        Self {
+            next: 1,
+            watermark: 0,
+            batch_counts: HashMap::new(),
+            pending_sequences: HashMap::new(),
+        }
+    }
+}
+
+/// Tracks delivery state across all messages.
+///
+/// Note on sequence overflow: Using u64 for sequence numbers provides ample headroom.
+/// At 100 messages per nanosecond, it would take ~5.85 years to overflow.
+/// For systems requiring longer uptime without restart, consider implementing
+/// sequence number wrapping with careful watermark handling.
+#[derive(Clone)]
+struct Tracker<B: Eq + Hash + Clone> {
+    state: Arc<Mutex<State<B>>>,
+}
+
+impl<B: Eq + Hash + Clone> Tracker<B> {
+    fn new() -> Self {
+        Self {
+            state: Arc::new(Mutex::new(State::default())),
+        }
+    }
+
+    fn guard(&self, batch: Option<B>) -> Guard<B> {
+        // Get state
+        let mut state = self.state.lock().unwrap();
+
+        // Get the next sequence
+        let sequence = state.next;
+        state.next += 1;
+
+        // Track this sequence as not yet delivered
+        state.pending_sequences.insert(sequence, false);
+
+        // Update batch count if provided
+        if let Some(batch) = &batch {
+            *state.batch_counts.entry(batch.clone()).or_insert(0) += 1;
+        }
+
+        Guard {
+            sequence,
+            tracker: self.state.clone(),
+
+            _batch: batch,
+        }
+    }
+}
+
+/// A sender that wraps `Sender` and tracks message delivery.
+#[derive(Clone)]
+pub struct Sender<T, B: Eq + Hash + Clone> {
+    inner: FutSender<Message<T, B>>,
```

Comment:
In consensus, I plan to use `view` as the batch (and prune a `view` only once all messages have been delivered).